libgit/repo.go

```diff
@@ -545,3 +571,167 @@ func DeleteRepo(
 		ctx, repoNode, normalizedRepoName, deletedReposNode,
 		normalizedRepoName+dirSuffix)
 }
+
+func renameRepoInConfigFile(
+	ctx context.Context, repoFS billy.Filesystem, newRepoName string) error {
+	// Assume lock file is already taken for both the old repo and the
+	// new one.
+	f, err := repoFS.OpenFile(kbfsConfigName, os.O_RDWR, 0600)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	buf, err := ioutil.ReadAll(f)
+	if err != nil {
+		return err
+	}
+	c, err := configFromBytes(buf)
+	if err != nil {
+		return err
+	}
+	c.Name = newRepoName
+	buf, err = c.toBytes()
+	if err != nil {
+		return err
+	}
+	_, err = f.Seek(0, io.SeekStart)
+	if err != nil {
+		return err
+	}
+	err = f.Truncate(0)
+	if err != nil {
+		return err
+	}
+	_, err = f.Write(buf)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// RenameRepo renames the repo from an old name to a new name.  It
+// leaves a symlink behind so that old remotes will continue to work.
+// The caller is responsible for syncing the FS and flushing the
+// journal, if desired.
+func RenameRepo(
+	ctx context.Context, config libkbfs.Config, tlfHandle *libkbfs.TlfHandle,
+	oldRepoName, newRepoName string) error {
+	kbfsOps := config.KBFSOps()
+	rootNode, _, err := kbfsOps.GetOrCreateRootNode(
+		ctx, tlfHandle, libkbfs.MasterBranch)
+	if err != nil {
+		return err
+	}
+	normalizedOldRepoName := normalizeRepoName(oldRepoName)
+	normalizedNewRepoName := normalizeRepoName(newRepoName)
+
+	repoNode, _, err := kbfsOps.Lookup(ctx, rootNode, kbfsRepoDir)
+	if err != nil {
+		return err
+	}
+
+	// Does the old repo definitely exist?
+	_, _, err = kbfsOps.Lookup(ctx, repoNode, normalizedOldRepoName)
+	if err != nil {
+		return err
+	}
+
+	if oldRepoName == newRepoName {
+		// The names are the same, nothing else to do.
+		return nil
+	}
+
+	fs, err := libfs.NewFS(ctx, config, tlfHandle, path.Join(kbfsRepoDir), "")
+	if err != nil {
+		return err
+	}
+
+	oldRepoFS, err := fs.Chroot(normalizedOldRepoName)
+	if err != nil {
+		return err
+	}
+
+	// Take locks in both repos during rename (same lock that's taken
+	// for new repo creation).
+	oldLockFile, err := takeConfigLock(
+		oldRepoFS.(*libfs.FS), tlfHandle, oldRepoName)
+	if err != nil {
+		return err
+	}
+	defer oldLockFile.Close()
+
+	if normalizedOldRepoName == normalizedNewRepoName {
+		// All we need to do is update the name in the config file,
+		// and the MD.
+		err = renameRepoInConfigFile(ctx, oldRepoFS, newRepoName)
+		if err != nil {
+			return err
+		}
+		return UpdateRepoMD(ctx, config, tlfHandle, oldRepoFS)
+	}
+
+	// Does the new repo not exist yet?
+	_, _, err = kbfsOps.Lookup(ctx, repoNode, normalizedNewRepoName)
+	switch errors.Cause(err).(type) {
+	case libkbfs.NoSuchNameError:
+		// The happy path.
+	case nil:
+		newRepoFS, err := fs.Chroot(normalizedNewRepoName)
+		if err != nil {
+			return err
+		}
+		return makeExistingRepoError(ctx, config, newRepoFS, newRepoName)
```

Comment:
What if the new repo location is a symlink, i.e. the location's repo was renamed to something else? It seems we should override the symlink here too, since one of scenarios where user wants to rename repos is they might want to reuse the name for something else.

Reply:
Oh yeah, I meant to do that and forgot, good catch.  Will work on that now.